toolchain/install/symlink_helpers.bzl

```
@@ -46,14 +46,22 @@ symlink_file = rule(
 
 def _symlink_filegroup_impl(ctx):
     prefix = ctx.attr.out_prefix
+    remove_prefix = ctx.attr.remove_prefix
 
     outputs = []
     for f in ctx.files.srcs:
         # We normalize the path to be package-relative in order to ensure
-        # consistent paths across possible repositories.
-        relative_path = f.short_path.removeprefix(f.owner.package)
+        # consistent paths across possible repositories. After experimenting,
+        # incrementally removing these path fragments, in this order,
+        # effectively handles the full permutation of files across different
+        # repositories and both generated and source files.
+        relative_path = f.path
+        relative_path = relative_path.removeprefix(f.root.path + "/")
+        relative_path = relative_path.removeprefix(f.owner.workspace_root + "/")
+        relative_path = relative_path.removeprefix(f.owner.package + "/")
+        relative_path = relative_path.removeprefix(remove_prefix)
```

Should this assert that it has the `remove_prefix`?

```suggestion
        if not relative_path.startswith(remove_prefix):
            fail("Missing `{0}` in `{1}`".format(remove_prefix, relative_path))
        relative_path = relative_path.removeprefix(remove_prefix)
```

---

No, we end up with some combination of these prefixes, but not all of them. We're using the fact that `removeprefix` is a no-op when the prefix is missing.

---

A couple things here:

1) If the "remove_prefix" is optional, maybe it should be "maybe_remove_prefix" or otherwise documented?

2) Why does "@llvm-project//clang:builtin_headers_gen" only _sometimes_ have a `staging/include/` prefix? Can this be documented?

---

Note, I'm especially confused about the latter because https://github.com/llvm/llvm-project/blob/main/utils/bazel/llvm-project-overlay/clang/BUILD.bazel#L1788 makes it look like it's always there.

---

Maybe you can list the specific environments you're supporting and the combinations in each? That might also help me understand what you're trying to address here, and might also assist with maintenance in the future.